### ISSUE TITLE
Document setup to run expectations?

### ISSUE THREAD

**Issue by Vaguery** [aws/aws-flow-ruby#94]
I was rolling up my sleeves to try to figure out the Ruby 2.2 problem in issue #88, so I created a clean Ruby 2.2.1 environment with `aws cli` and  `aws-flow-ruby` installed, configured it with the right access keys, and invoked `rspec -c -fd` to see what works out of the box.

The first four expectations seem to have hung (see below), and I had to interrupt them manually after a few minutes each.

I assume I'm missing something configuration-wise: Are these dependent on settings I don't have set in my config, or something else not documented in the specs themselves? I'm not seeing any callouts or comments saying "For these to work you need…"

What am I missing?

~~~ text
rspec -c -fd

AWS::Flow::Activities
Using test domain: rubyflow_test_domain
^C  ensures that a real activity will get scheduled (FAILED - 1)
^C  tests to see what two activities look like (FAILED - 2)
^C  tests to see that two subsequent activities are supported (FAILED - 3)
^C  tests a much larger workflow (FAILED - 4)
  Github issue 57
    ensures _options method returns an array of type ActivityType
    ensures the activity gets set with the right options
~~~

**Comment by pmohan6** [comment]
Really appreciate you digging into the issue.

I don't think this is a configuration issue. The reason why you see the tests hang is because -

The first test has the following lines -
```ruby
# spec/aws/decider/integration/activity_spec.rb
 62     worker.run_once
 63     activity_worker.run_once
 64     worker.run_once
 65     wait_for_execution(workflow_execution)
```
Here, the workflow worker will run once - i.e. poll for a task and make a decision. But because of the issue we are facing, the worker can't find the workflow definition.
```
#### Workflow Worker log - /tmp/AWS-Flow-WorkflowWorker
ERROR -- : Error in the poller, #<RuntimeError: No workflow definition for <AWS::SimpleWorkflow::WorkflowType domain:rubyflow_test_domain-2015-03-06-01 name:BasicWorkflow.start version:1.0>>
```
Hence, the workflow ends up timing out. Execution start to close timeout value is set to 10 minutes on this workflow, hence it takes a while for the test to fail. We should probably adjust this value to be much stricter.

Can you check if you are able to run the tests with other versions of ruby?

**Comment by Vaguery** [comment]
More tests pass in 2.1.5, but there are still a few troubling spots: logging errors, a lot more passing tests but with failed workers (on the SWF pane) which I assume are intentional checks... hard to say.

To be honest, I'm evaluating the quality of the service for a personal project, and have no funding. I would hope (and prefer) that the people who are actually working at Amazon _tell_ us that the tests are passing, and under which Ruby interpreters. Like using a test harness and stuff like professionals use when they have clouds and stuff?

**Comment by pmohan6** [comment]
That's a fair point. We will work on improving our documentation around our tests. As of now, all tests pass with ruby 2.1.5. (There are few non-deterministic integration tests which might need to be re-run). 

Btw, workflow failures/timeouts on the SWF console might be intentional tests checking failure scenarios.

**Comment by Vaguery** [comment]
Is there any testing mode that doesn't actually rack up money? As in unit tests _qua_ unit tests, rather than full-stack round-trip integration tests?

**Comment by Vaguery** [comment]
Just to clarify something else I'm trying to figure out, what exactly is [going on here?](https://github.com/aws/aws-flow-ruby/blob/master/aws-flow/lib/aws/decider/async_decider.rb#L56)

**Comment by Vaguery** [comment]
@pmohan6 They don't all pass on my clean 2.1.5 install, and there are apparently not even any rspec tags on the integration tests to differentiate them from the unit tests. There are several places where the code clearly has been half-filled in but not tested, as well, so I'm not sure what relation the tests have _to_ the codebase if any.

**Comment by pmohan6** [comment]
Unfortunately, we don't have a testing mode in the service. You can run the framework unit tests using ```rake unit_tests``` and all tests using ```rake all_tests```.

Admittedly, unit test coverage in aws-flow is not the best. During our initial release, most of our tests validated that features worked as expected. Since then, we have added a lot more coverage but there is still more work to be done. We are working to improve our test quality.

I'll take a look at the line that you referenced above today.

**Comment by pmohan6** [comment]
We've released a fix for the ruby 2.2.2 issue. Please let us know if you see any further issues. Closing this for now. Thanks for digging into the issue with us!

We will update our testing documentation in the meantime.
